feat: adding Falcon auth

Dominik1999 opened the pull request:
Closes #293 

Since https://github.com/0xPolygonMiden/miden-vm/issues/951 is closed now, we can implement the real transaction scheme into the `auth_tx_rpo_falcon512` procedure.

frisitano commented:
I've had a chance to take a look at this and I believe I've identified the root cause.  As part of the `get_output_notes_hash` procedure we loop over created notes and compute the vault hash of each note, whilst computing this vault hash we also insert the assets from the note into the output vault.  This is to allow us to compute asset invariant assertions. When we invoke this procedure prior to the epilogue the output vault hasn't been created (copied from the input vault) yet and as such we receive a `MerkleStoreLookupFailed` error. 

We have two options of how we can address this, we can move the logic for inserting created note assets into the output vault into a separate procedure or we can leave the logic where it is and put it behind a boolean flag.  If the `get_output_notes_hash` is called via the kernel api then the boolean flag will be set to false (i.e. skip output vault insertion logic) and if it is called from the epilogue then it will be set to true.

frisitano commented:
After rebasing #297, all tests are passing.

Dominik1999 commented:
Nice, it works now. 

@frisitano I changed the `TransactionExecutor` to receive the tupel `(pub_key, keypair)` that is necessary for the falcon verification. 

I changed all integration tests to test the new signature scheme now.

frisitano commented:
I'm going to work on implementing https://github.com/0xPolygonMiden/miden-base/issues/300 which we can then use in this PR.

frisitano commented:
I've opened an initial draft for the `TransactionScript` object integration - https://github.com/0xPolygonMiden/miden-base/pull/301.  I've requested a review from @bobbinth to evaluate if it is looks ok.  If so then I will productionise it.

Dominik1999 commented:
Sounds great, maybe you can give the rest an initial review? Then I can finalize this PR more quickly

frisitano commented:
Sure can you rebase main please.

Dominik1999 commented:
> Sure can you rebase main please.

Should I simply rebase on #301 ?

frisitano commented:
> > Sure can you rebase main please.
> 
> Should I simply rebase on #301 ?

Yes please, but only after I have completed changes.

Dominik1999 commented:
> > > Sure can you rebase main please.
> > 
> > 
> > Should I simply rebase on #301 ?
> 
> Yes please, but only after I have completed changes.

Ready?

frisitano commented:
> > > > Sure can you rebase main please.
> > > 
> > > 
> > > Should I simply rebase on #301 ?
> > 
> > 
> > Yes please, but only after I have completed changes.
> 
> Ready?

I've submitted it for review so there may be some changes requested from that. I think we should wait until that is completed and merged to avoid duplicating the amount of work we have to do here.

frisitano commented:
If you could rebase main please and then I will review this @Dominik1999.

Dominik1999 commented:
> If you could rebase main please and then I will review this @Dominik1999.

ok done, you can start the review

frisitano commented on miden-lib/asm/eoa/basic.masm:
```diff
@@ -1,32 +1,47 @@
+# Slot in account storage at which the public key is stored.
+const.PUBLIC_KEY_SLOT=0
+
+#! Authenticate a transaction using Falcon 
```
minor nit: "Authenticate a transaction using Falcon" -> "Authenticate a transaction using the Falcon signature scheme"

frisitano commented on miden-lib/asm/eoa/basic.masm:
```diff
@@ -1,32 +1,47 @@
-    #exec.account::get_item
-    # [PUB_KEY, M]
-    # TODO: sign the message using FALCON
+    # Get commitments to consumed and created notes, current nonce, and ID
```
optional nit: it may be preferable to split this comment up and comment the lines below according to the entity being fetched such that each operation has a comment.

frisitano commented on miden-lib/asm/eoa/basic.masm:
```diff
@@ -1,32 +1,47 @@
+
+    # Compute the message to be signed
+    # M = h(OUTPUT_NOTES_HASH, h(INPUT_NOTES_HASH, h(0, 0, 0, account_id, 0, 0, 0, nonce)))
+    hmerge hmerge hmerge
```
question: Is there a reason we use 2 to 1 hashing here instead of using a sequential hash over all elements which would be more efficient?

Dominik1999 replied:
Isn't the 2-to-1 hash more efficient? 

https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/cryptographic_operations.html

Or maybe I don't get it. Which instruction would you use?

frisitano replied:
The `hperm` operation for sequential hashing.  It would look something like this (please review yourself and adjust accordingly to ensure correctness):

```
# prepare hasher capacity elements
padw 

# prepare first hasher rate word
exec.account::get_nonce push.0.0.0

# prepare second hasher rate word
exec.account::get_id push.0.0.0

# apply first permutation
hperm

# drop rate elements
dropw dropw

# prepare third hasher rate word
exec.tx::get_input_notes_hash

# prepare fourth hasher rate word
exec.tx::get_output_notes_hash

# apply second hasher permutation
hperm

# extract the digest (message)
dropw swapw dropw
```

This will give a different result to the `hmerge` method but it is a more efficient way of computing a message.

Dominik1999 replied:
`hmerge` was used as it was the easiest way to construct how the message M was specified. 

`M = h(OUTPUT_NOTES_HASH, h(INPUT_NOTES_HASH, h(0, 0, 0, account_id, 0, 0, 0, nonce)))`

However, if the other method is more efficient. Should I open an issue and we discuss this?

frisitano replied:
Yes please lets leave it as it is and open an issue to discuss this.

frisitano commented on miden-tx/tests/common/mod.rs:
```diff
@@ -78,3 +87,61 @@ impl DataStore for MockDataStore {
+    let keypair: KeyPair = KeyPair::new().unwrap();
+
+    let pk: Word = keypair.public_key().into();
+    let pk_sk_bytes = keypair.to_bytes();
+    let to_adv_map = pk_sk_bytes.iter().map(|a| Felt::new(*a as u64)).collect::<Vec<Felt>>();
```
We should consider implementing:

```
impl From<KeyPair> for [Word; 2] {
...
}
```

To avoid having to serialize to bytes and then cast to felts.  This should make the interface cleaner. Lets create an issue in crypto to discuss this.  Not required for this PR.

frisitano commented on miden-tx/tests/common/mod.rs:
```diff
@@ -78,3 +87,61 @@ impl DataStore for MockDataStore {
+
+// HELPER FUNCTIONS
+// ================================================================================================
+pub fn get_new_key_pair_with_advice_map() -> (KeyPair, ([Felt; 4], Vec<Felt>)) {
```
It looks like we some redundancy here. We return the `KeyPair` which contains the `PublicKey` and secret key. We also return `(PublicKey as Word, Vec[PublicKey as Word, SecretKey as Word]))`. So the public key is contained in 3 of the returned components and the secret key in two. You could just return `(Word, Word)`. One word for the public key and one word for the private key.

This could then be invoked like:

`let (pk, sk) = get_new_keypair();`

When you define the entry for the `TransactionScript` inputs it would be constructed as `(pk, [pk, sk].concat())`.

I think this is a bit cleaner.

Dominik1999 replied:
That makes a lot of sense

Dominik1999 replied:
I only don't get this 

```
    let pk_sk_felts: Vec<Felt> = pk_sk_bytes.iter().map(|a| Felt::new(*a as u64)).collect::<Vec<Felt>>();
```

into a `Word`. Maybe we should return `(Word, Vec<Felt>)` whereas `Vec<Felt>` represents the `sk`?

frisitano replied:
You should be able to do it as:

```
    let pk_sk_felts = pk_sk_bytes.iter().map(|a| Felt::new(*a as u64)).collect::<Vec<Felt>>();
    let pk = Word::try_from(&pk_sk_felts[0..4]).unwrap();
    let sk = Word::try_from(&pk_sk_felts[4..]).unwrap();
    (pk, sk)
```

Dominik1999 replied:
Hmm using 

```
pub fn get_new_key_pair_with_advice_map() -> (Word, Word) {
    let keypair: KeyPair = KeyPair::new().unwrap();
    let pk_sk_bytes = keypair.to_bytes();
    let _pk_sk_felts = pk_sk_bytes.iter().map(|a| Felt::new(*a as u64)).collect::<Vec<Felt>>();
        let pk_sk_felts = pk_sk_bytes.iter().map(|a| Felt::new(*a as u64)).collect::<Vec<Felt>>();
    let pk = Word::try_from(&pk_sk_felts[0..4]).unwrap();
    let sk = Word::try_from(&pk_sk_felts[4..]).unwrap();

    (pk, sk)
}
```

results in 

```
thread 'test_faucet_contract_mint_fungible_asset_succeeds' panicked at miden-tx/tests/common/mod.rs:95:48:
called `Result::unwrap()` on an `Err` value: TryFromSliceError(())
```

whereas line 95 

```
    let sk = Word::try_from(&pk_sk_felts[4..]).unwrap();
```

Might it be that the sk part of this is more than what fits into a `Word`?

frisitano replied:
Ahh yess - quite possibly.  I've just taken a look at the key configuration and it looks like theres an expanded public key and a public key commitment so what you have at the moment is perfect. In effect you are returning

```
(pk_commitment, Vec[expanded_public_key, secret_key])
```

Lets leave it as is.  Additionally https://github.com/0xPolygonMiden/crypto/issues/217 doesn't make sense anymore because the expanded pk, sk pair wont fit into words.

Dominik1999 replied:
All good, then I think we are good to go

frisitano commented on miden-tx/tests/test_miden_faucet_contract.rs:
```diff
@@ -246,6 +265,7 @@ fn test_faucet_contract_creation() {
 }
 
 fn get_faucet_account_with_max_supply_and_total_issuance(
+    public_key: PublicKey,
```
If we implement the change for key pair creation then this can be changed to `public_key: Word`

frisitano commented on miden-tx/tests/common/mod.rs:
```diff
@@ -78,3 +87,61 @@ impl DataStore for MockDataStore {
+#[allow(dead_code)]
+pub fn get_account_with_default_account_code(
+    account_id: AccountId,
+    public_key: PublicKey,
```
If we implement the changes described for key pair creation then this can be `public_key: Word`.

frisitano reviewed (changes requested):
Looks good, thanks!  The logic looks correct.  I've added some questions and suggestions to improve code hygiene. Let me know what you think.

frisitano reviewed (approved):
Looks good, thanks!
